cmd/kraft/pkg/push/push.go

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
+// Licensed under the BSD-3-Clause License (the "License").
+// You may not use this file except in compliance with the License.
+package push
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/spf13/cobra"
+
+	"kraftkit.sh/cmdfactory"
+	"kraftkit.sh/config"
+	"kraftkit.sh/log"
+	"kraftkit.sh/pack"
+	"kraftkit.sh/packmanager"
+	"kraftkit.sh/tui/paraprogress"
+	"kraftkit.sh/unikraft/app"
+)
+
+type Push struct {
+	Format string `local:"true" long:"as" short:"M" usage:"Force the packaging despite possible conflicts" default:"auto"`
+}
+
+func New() *cobra.Command {
+	cmd, err := cmdfactory.New(&Push{}, cobra.Command{
+		Short:   "Push a Unikraft unikernel package to registry",
+		Use:     "push [FLAGS] [PACKAGE]",
+		Aliases: []string{"ph"},
+		Long: heredoc.Doc(`
+			Push a Unikraft unikernel, component microlibrary to a remote location
+		`),
+		Example: heredoc.Doc(`
+			# Push the image for a project in the current directory
+			$ kraft pkg push
+
+			# Push the image for a project at a path with tag latest
+			$ kraft pkg push /path/to/app
+
+			# Push the image with a given name
+			$ kraft pkg push unikraft.org/helloworld:latest`),
+		Annotations: map[string]string{
+			cmdfactory.AnnotationHelpGroup: "pkg",
+		},
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	return cmd
+}
+
+func (opts *Push) Pre(cmd *cobra.Command, _ []string) error {
+	ctx := cmd.Context()
+	pm, err := packmanager.NewUmbrellaManager(ctx)
+	if err != nil {
+		return err
+	}
+
+	cmd.SetContext(packmanager.WithPackageManager(ctx, pm))
+
+	return nil
+}
+
+func (opts *Push) Run(cmd *cobra.Command, args []string) error {
+	var err error
+	var workdir string
+
+	if len(args) == 0 {
+		workdir, err = os.Getwd()
+		if err != nil {
+			return err
+		}
+	} else if f, err := os.Stat(args[0]); err == nil && f.IsDir() {
+		workdir = args[0]
+	} else {
+		workdir = ""
+	}
+
+	ctx := cmd.Context()
+	norender := log.LoggerTypeFromString(config.G[config.KraftKit](ctx).Log.Type) != log.FANCY
+	ref := ""
+	if workdir != "" {
+		// Read the kraft yaml specification and get the target name
+		project, err := app.NewProjectFromOptions(
+			ctx,
+			app.WithProjectWorkdir(workdir),
+			app.WithProjectDefaultKraftfiles(),
+		)
+		if err != nil {
+			return err
+		}
+
+		// Get the target name
+		ref = project.Name()
+	} else {
+		// Argument is a reference name
+		ref = args[0]
+
+		// Check if a tag was specified
+		if (strings.ContainsRune(ref, ':') && strings.Split(ref, ":")[1] == "") ||
+			strings.Count(ref, ":") > 1 {
+			return errors.New("invalid tag specified")
+		}
```

Comment:
The use of `:` is package-manager specific. We can alleviate this check by simply passing `args[0]` into the `Name` attribute of `CatalogQuery`. This code be removed.